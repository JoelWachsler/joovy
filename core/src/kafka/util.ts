import { Kafka } from 'kafkajs'
import { defer, from, mapTo, mergeMap, Observable, tap } from 'rxjs'
import initConfig from '../config'
import logger from '../logger'
import { KMessage } from './kafka'

let kafka: Kafka

/**
 * Internal implementation detail, do not use.
 */
export const createKafkaClient = () => {
  if (!kafka) {
    const kafkaConfig = initConfig().kafka

    kafka = new Kafka({
      clientId: kafkaConfig.clientId,
      brokers: kafkaConfig.brokers,
    })
  }

  return kafka
}

type CreateConsumerArgs = {
  topic: string
  groupId: string
}

export const createConsumer = <T extends KMessage = KMessage>(args: CreateConsumerArgs) => {
  const { groupId, topic } = args

  const create = defer(() => {
    logger.info(`Creating consumer with the following args: ${JSON.stringify(args)}`)
    const consumer = createKafkaClient().consumer({ groupId })

    return from(consumer.connect()).pipe(mapTo(consumer))
  })

  const connectToTopic = create.pipe(mergeMap(consumer => {
    logger.info(`Subscribing to topic: ${topic}...`)
    return from(consumer.subscribe({ topic, fromBeginning: true })).pipe(
      tap(() => logger.info(`Subscribed to topic: ${topic}!`)),
      mapTo(consumer),
    )
  }))

  return connectToTopic.pipe(mergeMap(consumer => {
    return new Observable<T>(subscribe => {
      consumer.run({
        async eachMessage({ message, topic }) {
          const msgAsString = message.value?.toString()
          if (!msgAsString) {
            logger.error(`Invalid message: ${message} for topic: ${topic}`)
            return
          }

          logger.info(`New incoming message: ${JSON.stringify(msgAsString)} on topic: ${topic}`)

          subscribe.next(JSON.parse(msgAsString))
        },
      })
    })
  }))
}