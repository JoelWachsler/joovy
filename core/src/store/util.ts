import { defer } from 'rxjs'
import JEvent from '../jevent/JEvent'

const messageStore = new Map<string, Map<string, unknown>>()

export const getOrCreateObjectStore = (id: string) => {
  let store = messageStore.get(id)
  if (!store) {
    store = new Map()
    messageStore.set(id, store)
  }

  return store
}

export const removeObjectStore = (event: JEvent) => {
  return defer(() => {
    const channelId = event.message.channelId
    messageStore.delete(channelId)

    return event.result({ store: { event: 'removed', id: channelId } })
  })
}
