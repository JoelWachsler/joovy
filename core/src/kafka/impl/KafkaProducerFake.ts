import { defer, Observable } from 'rxjs'
import { Result } from '../../jevent/Result'
import { KafkaProducer, SendArgs } from '../kafka'

export class KafkaProducerFake implements KafkaProducer {

  send({ message, topic, event }: SendArgs): Observable<Result> {
    return defer(() => event.result({ kafka: { message, topic } }))
  }
}
