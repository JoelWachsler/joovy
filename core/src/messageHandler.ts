import { Client } from 'discord.js'
import { catchError, filter, mergeMap, Observable } from 'rxjs'
import { handle } from './commands/command'
import { errorHandler } from './errorHandler'
import JEvent, { fromMessageKey } from './jevent/JEvent'
import { EmptyResult, Result } from './jevent/Result'
import { createConsumer, createProducer, KMessage, Topics } from './kafka/kafka'
import { logResult } from './resultLogger'

export const sendMessageEvent = (event: Observable<JEvent>): Observable<Result> => {
  return createProducer().pipe(mergeMap(producer => {
    return event.pipe(mergeMap(event => {
      return producer.send({
        event,
        message: createMessageEvent(event),
        topic: Topics.NewMessage,
      })
    }))
  }))
}

export const handleMessageEvents = (client: Client) => {
  const messageHandling = createConsumer<MessageEvent>({ groupId: 'joovy-message-handler', topic: Topics.NewMessage }).pipe(
    mergeMap(msg => fromMessageKey(client, msg.meta.messageKey)),
    handleMessage,
  )

  logResult('handleMessageEvents', messageHandling)
}

export const handleMessage = (event: Observable<JEvent>): Observable<Result> => {
  return event.pipe(
    mergeMap(event => {
      const message = event.message
      if (message.author.bot) {
        return event.result({ ignored: `${message.content} was sent by a bot` })
      } else if (!message.content.startsWith('/')) {
        return event.result({ ignored: `${message.content} does not start with a slash` })
      } else {
        return handle(event).pipe(
          filter(r => !(r instanceof EmptyResult)),
          catchError(err => errorHandler(event, err)),
        )
      }
    }),
  )
}

const createMessageEvent = (event: JEvent): MessageEvent => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'message-event',
      version: '1',
    },
    content: event.message.content,
  }
}

type MessageEvent = KMessage<'message-event', '1'> & { content: string }
