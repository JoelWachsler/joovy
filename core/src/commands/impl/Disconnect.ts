import { Client } from 'discord.js'
import { concat, mergeMap, Observable } from 'rxjs'
import JEvent, { fromMessageKey } from '../../jevent/JEvent'
import { Result } from '../../jevent/Result'
import { KMessage, Topics } from '../../kafka/kafka'
import { createConsumer } from '../../kafka/util'
import { removePlaylist } from '../../playlist/Playlist'
import { logResult } from '../../resultLogger'
import { removeObjectStore } from '../../store/util'
import ArgParser from '../ArgParser'
import Command from '../command'

export default class Disconnect implements Command {
  argument = ArgParser.create('disconnect')
  helpText = 'Disconnects the bot from the current channel.'

  handleMessage(event: JEvent): Observable<Result> {
    return event.factory.kafkaProducer().pipe(mergeMap(producer => {
      return producer.send({
        topic: Topics.Disconnect,
        event,
        message: createDisconnectMessage(event),
      })
    }))
  }
}

export const handleDisconnects = (client: Client) => {
  const disconnectHandler = createConsumer<DisconnectMessage>({ groupId: 'joovy-disconnect-handler', topic: Topics.Disconnect }).pipe(
    mergeMap(message => fromMessageKey(client, message.meta.messageKey)),
    disconnectFromChannel,
  )

  logResult('disconnect', disconnectHandler)
}

export const disconnectFromChannel = (event: Observable<JEvent>) => {
  return event.pipe(
    mergeMap(event => {
      const playlistRemoval = removePlaylist(event)
      const removeChannelStore = removeObjectStore(event)
      const sendByeMessage = event.sendMessage('Bye!')

      return concat(playlistRemoval, removeChannelStore, sendByeMessage)
    }),
  )
}

type DisconnectMessage = KMessage<'disconnect', '1'>

export const createDisconnectMessage = (event: JEvent): DisconnectMessage => {
  return {
    meta: {
      messageKey: event.message.messageKey,
      name: 'disconnect',
      version: '1',
    },
  }
}
