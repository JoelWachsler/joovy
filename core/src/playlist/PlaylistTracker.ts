import { catchError, concat, defaultIfEmpty, map, mapTo, mergeAll, mergeMap, Observable, of } from 'rxjs'
import { errorHandler } from '../errorHandler'
import JEvent from '../jevent/JEvent'
import Track from '../player/Track'
import { getOrCreateStore } from '../store/impl/LevelStore'
import { StringStore } from '../store/Store'
import { Playlist } from './Playlist'

export const trackPlaylist = (event: JEvent, playlist: Playlist) => {
  const events = event.store.persistentString.pipe(
    map(store => createRepositories(store)),
    mergeMap(repos => {
      const playlistId = event.factory.uuid()
      const { channelId } = event.message

      const createPlaylist = repos.channel.getOrCreate(event, channelId).pipe(
        mergeMap(channel => {
          channel.playlists = [...channel.playlists, playlistId]
          return repos.channel.put(channelId, channel)
        }),
      )

      const tracking = concat(
        createPlaylist,
        playlist.internalCurrentQueue.pipe(
          mergeMap(currentQueue => {
            return repos.playlist.getOrCreate(event, playlistId).pipe(
              mergeMap(queue => {
                queue.tracks = currentQueue.map(track => trackToTrackV1(track))
                return repos.playlist.put(playlistId, queue)
              }),
            )
          }),
        ),
      )

      return tracking.pipe(
        mergeMap(e => event.complexResult({ result: TRACKING_EVENT, item: e })),
      )
    }),
  )

  return events.pipe(
    // prevent whole play-session from crashing if something goes bad here
    catchError(err => errorHandler(event, err)),
  )
}

export const TRACKING_EVENT = { playlist: 'trackingEvent' }

abstract class Repository<Entity extends Meta, ID extends string = string> {
  constructor(private store: StringStore) { }

  get(id: ID): Observable<Entity> {
    return this.store.get(id).pipe(map(e => JSON.parse(e) as unknown as Entity))
  }

  put(id: ID, entity: Entity): Observable<Entity> {
    return this.store.put(id, JSON.stringify(entity)).pipe(mapTo(entity))
  }

  getOrCreate(event: JEvent, id: ID): Observable<Entity> {
    return this.get(id).pipe(
      map(e => of(e)),
      defaultIfEmpty(this.put(id, this.defaultEntity(event))),
      mergeAll(),
    )
  }

  abstract defaultEntity(event: JEvent): Entity
}

class ChannelRepository extends Repository<ChannelV1> {
  defaultEntity(): ChannelV1 {
    return {
      meta: {
        name: Type.Channel,
        version: Version.v1,
      },
      playlists: [],
    }
  }
}

class PlaylistRepository extends Repository<PlaylistV1> {
  defaultEntity(event: JEvent): PlaylistV1 {
    return {
      meta: {
        name: Type.Playlist,
        version: Version.v1,
      },
      date: event.timestamp,
      tracks: [],
    }
  }
}

const createRepositories = (store: StringStore) => {
  return {
    channel: new ChannelRepository(store),
    playlist: new PlaylistRepository(store),
  }
}

export const dumpV1 = (): Promise<DumpV1> => {
  return new Promise(resolve => {
    const dump: DumpV1 = {
      channels: [],
      playlists: [],
    }

    getOrCreateStore().pipe(mergeMap(store => store.dump())).subscribe({
      next: keyValue => {
        const { key, value } = keyValue
        const item = JSON.parse(value.toString()) as Meta

        if (isChannel(item)) {
          dump.channels.push({
            key: key.toString(),
            channel: item,
          })
        } else if (isPlaylist(item)) {
          dump.playlists.push(item)
        } else {
          throw Error(`Unknown object: ${item}`)
        }
      },
      complete: () => resolve(dump),
    })
  })
}

const trackToTrackV1 = (track: Track): TrackV1 => {
  return {
    meta: {
      name: Type.Track,
      version: Version.v1,
    },
    author: {
      meta: {
        name: Type.Author,
        version: Version.v1,
      },
      id: track.author.id,
      username: track.author.username,
    },
    name: track.name,
    link: track.link,
  }
}

const isChannel = (item: Meta): item is ChannelV1 => {
  return item.meta.name === Type.Channel
}

const isPlaylist = (item: Meta): item is PlaylistV1 => {
  return item.meta.name === Type.Playlist
}

interface DumpV1 {
  channels: ChannelDumpV1[]
  playlists: PlaylistV1[]
}

interface ChannelDumpV1 {
  key: string
  channel: ChannelV1
}

interface ChannelV1 extends Meta<Type.Channel, Version.v1> {
  playlists: PlaylistIdentifier[]
}

type PlaylistIdentifier = string

interface PlaylistV1 extends Meta<Type.Playlist, Version.v1> {
  date: number
  tracks: TrackV1[]
}

interface TrackV1 extends Meta<Type.Track, Version.v1> {
  author: AuthorV1
  name: string
  link: string
}

interface AuthorV1 extends Meta<Type.Author, Version.v1> {
  id: string
  username: string
}

interface Meta<Name = string, Version = string> {
  meta: {
    name: Name
    version: Version
  }
}

enum Type {
  Channel = 'channel',
  Playlist = 'playlist',
  Track = 'track',
  Author = 'author',
}

enum Version {
  v1 = '1',
}
