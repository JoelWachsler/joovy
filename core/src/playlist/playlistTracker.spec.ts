import { map, merge, Observable } from 'rxjs'
import { disconnectFromChannel } from '../commands/impl/Disconnect'
import { handleMessage } from '../messageHandler'
import { createTestEvent, hot, sandbox } from '../test/util'
import { TRACKING_EVENT } from './PlaylistTracker'

test('should create tracker for playlist', () => {
  const play = createTestEvent({ content: '/play test' })

  const messages = sandbox.getMessages(handleTrackingMessages(hot('a', { a: play })))

  expect(messages).toMatchSnapshot()
})

test('should track multiple tracks in the same playlist', () => {
  const play = createTestEvent({ content: '/play test' })
  const play2 = createTestEvent({ content: '/play test2' })

  const messages = sandbox.getMessages(handleTrackingMessages(hot('ab', { a: play, b: play2 })))

  expect(messages).toMatchSnapshot()
})

test('should differentiate tracking between playlists', () => {
  const play = createTestEvent({ content: '/play test' })
  const disconnect = createTestEvent({ content: '/disconnect' })

  const messages = handleTrackingMessages(hot('a-a', { a: play }))
  const messages2 = disconnectFromChannel(hot('-b-', { b: disconnect })).pipe(
    map(msg => msg.result),
  )

  expect(sandbox.getMessages(merge(messages, messages2))).toMatchSnapshot()
})

const handleTrackingMessages = (source: Observable<any>) => handleMessage(source).pipe(
  map(msg => {
    // include the actual item when matching tracking events
    if (msg.result === TRACKING_EVENT) {
      return msg
    } else {
      return msg.result
    }
  }),
)
