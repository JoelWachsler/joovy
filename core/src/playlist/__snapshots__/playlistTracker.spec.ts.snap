// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create tracker for playlist 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "meta": Object {
            "name": "channel",
            "version": "1",
          },
          "playlists": Array [
            "totally-random-uuid-0",
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
            },
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should differentiate tracking between playlists 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "meta": Object {
            "name": "channel",
            "version": "1",
          },
          "playlists": Array [
            "totally-random-uuid-0",
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
            },
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "disconnected",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "store": Object {
          "event": "removed",
          "id": "testChannelId",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Bye!\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "meta": Object {
            "name": "channel",
            "version": "1",
          },
          "playlists": Array [
            "totally-random-uuid-0",
            "totally-random-uuid-1",
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
            },
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 2,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 7,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 7,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;

exports[`should track multiple tracks in the same playlist 1`] = `
Array [
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "created",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "meta": Object {
            "name": "channel",
            "version": "1",
          },
          "playlists": Array [
            "totally-random-uuid-0",
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
            },
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test (1:10)](https://test.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 0,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test (1:10)](https://test.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "commandCalled": "/play",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "playlist": "found",
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "event": EventFake {
          "message": JMessageFake {
            "_reactions": Array [],
            "author": Object {
              "bot": false,
              "id": "testAuthorId",
              "username": "testuser",
            },
            "channelId": "testChannelId",
            "content": "/play test",
          },
          "timestamp": 949363200000,
        },
        "item": Object {
          "date": 949363200000,
          "meta": Object {
            "name": "playlist",
            "version": "1",
          },
          "tracks": Array [
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test (1:10)](https://test.com) [<@testAuthorId>]",
            },
            Object {
              "author": Object {
                "id": "testAuthorId",
                "meta": Object {
                  "name": "author",
                  "version": "1",
                },
                "username": "testuser",
              },
              "link": "https://test2.com",
              "meta": Object {
                "name": "track",
                "version": "1",
              },
              "name": "[test2 (1:10)](https://test2.com) [<@testAuthorId>]",
            },
          ],
        },
        "result": Object {
          "playlist": "trackingEvent",
        },
      },
    },
  },
  TestMessageValue {
    "frame": 1,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"[test2 (1:10)](https://test2.com) [<@testAuthorId>] has been added to the queue\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 5,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"Now playing: [test2 (1:10)](https://test2.com) [<@testAuthorId>]\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "player": "idle",
      },
    },
  },
  TestMessageValue {
    "frame": 10,
    "notification": Object {
      "kind": "N",
      "value": Object {
        "messageSent": "{
  \\"title\\": null,
  \\"type\\": \\"rich\\",
  \\"description\\": \\"End of playlist, will disconnect in 5 minutes.\\",
  \\"url\\": null,
  \\"timestamp\\": null,
  \\"color\\": null,
  \\"fields\\": [],
  \\"thumbnail\\": null,
  \\"image\\": null,
  \\"author\\": null,
  \\"footer\\": null
}",
      },
    },
  },
]
`;
